### PR TITLE
BC-589 - BC-654 - Change auto deployment to a reusable workflow

### PR DESCRIPTION
**Pull request by alriese**
# Description

## Links to Tickets or other pull requests

- hpi-schul-cloud/dof_app_deploy#57
- hpi-schul-cloud/schulcloud-server/pull/3169
- https://github.com/hpi-schul-cloud/schulcloud-client/pull/2713
- https://github.com/hpi-schul-cloud/nuxt-client/pull/1957
- https://github.com/hpi-schul-cloud/superhero-dashboard/pull/137
- https://github.com/hpi-schul-cloud/antivirus_check_service/pull/13

## Approval for review

- [x] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### Link to Definiton of Done

More and detailed information on the _definition of done_ can be found [on Confluence](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762)

### CODE REVIEW

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: test image exists
+        run: |
+         mkdir -p ~/.docker
+         echo '{"experimental": "enabled"}' >> ~/.docker/config.json
+         echo "IMAGE_EXISTS=$(docker manifest inspect ghcr.io/${{ github.repository }}:${{ github.sha }} > /dev/null && echo 1 || echo 0)" >> $GITHUB_ENV
+
```
Comment (virgilchiriac) [inline .github/workflows/push.yml:36]: could add docker buildx also
```suggestion
- name: Set up Docker Buildx
        if: ${{ env.IMAGE_EXISTS == 0 }}
        uses: docker/setup-buildx-action@v1
```

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Build and push ${{ github.repository }}
```
Comment (virgilchiriac) [inline .github/workflows/push.yml:37]:
```suggestion
      - name: Build and push ${{ github.repository }}
        if: ${{ env.IMAGE_EXISTS == 0 }}
```

File: .github/workflows/push.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./Dockerfile
+          platforms: linux/amd64
+          push: true
+          tags: ghcr.io/${{ github.repository }}:${{ github.sha }}
+          labels: ${{ steps.docker_meta_img.outputs.labels }}
+
```
Comment (virgilchiriac) [inline .github/workflows/push.yml:46]: branch name job missing
```suggestion
  branch_name:
    runs-on: ubuntu-latest
    outputs:
      branch: ${{ steps.extract_branch.outputs.branch }}
    steps:
      - name: Extract branch name
        shell: bash
        run: echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
        id: extract_branch
```

**Review by virgilchiriac — CHANGES_REQUESTED** [review]

File: .github/workflows/push.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
+---
+name: Build and push Docker Image
+
+on:
+  push:
+    branches-ignore:
+      - dependabot/**
+
```
Comment (mamutmk5) [inline .github/workflows/push.yml:8]: If you change the workflow to this way please remove Please remove https://github.com/hpi-schul-cloud/schulcloud-calendar/blob/BC-654/.github/workflows/main.yml

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

**Review by mamutmk5 — APPROVED** [review]

**Review by virgilchiriac — APPROVED** [review]

**Pull request merged**